File: baremetal/generators/software_models/vpu/lane_boxes/fp8_pack.py
"""lane_boxes/fp8_pack.py — funct model of FP8Pack.scala.

`FP8Pack` is the only **2 → 1 phased** lane_box on the BF16-side of the
engine. Two consecutive `req.valid` pulses (16 BF16 lanes each) feed
the phased state machine; one `resp.valid` pulse is emitted carrying a
`Vec(16, UInt(16))` packed row (32 FP8 bytes total, two per slot).

Cycle accounting (mirrors `FP8Pack.scala:101-129`):

    cycle K   : req=row_low,  !phase  → phase=true,  lowHalf=fp8(row_low),
                                        respValidReg=false
    cycle K+1 : req=row_high, phase=true → phase=false, respValidReg=true,
                                           respBitsReg=Cat(lowHalf, fp8(row_high))
    cycle K+2 : io.resp.valid=1, io.resp.bits.result=respBitsReg

The output is **registered** (`respValidReg`, `respBitsReg`) so it
appears one cycle AFTER the second pulse, not on the same cycle. The
phase register holds across idle cycles — `req.valid=false` clears
`respValidReg` but does NOT touch `phase` or `lowHalf`.

Output layout per `FP8Pack.scala:122-125`:

    result(j)     = Cat(lowHalf (2j+1), lowHalf (2j))   for j = 0..7
    result(j + 8) = Cat(fp8Bytes(2j+1), fp8Bytes(2j))   for j = 0..7

i.e. inside each 16-bit slot, byte 2j sits in bits[7:0] and byte 2j+1
sits in bits[15:8] (low byte first within the slot).

Per-byte BF16 → E4M3 conversion lives in `fp8_e4m3.bf16_to_e4m3_byte`,
already validated against `RVFP8PackTest.goldenFp8ByteFromBf16`.

Use `convert_row(req)` for a pure per-row BF16 → FP8 conversion (16
bytes out, no latching). Use `pack_two_rows(req_low, req_high)` for
the pure 2-row math without any phase state. Use `step("fp8pack", req)`
for cycle-accurate replay.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from .. import fp8_e4m3 as fp8
from ..vector_params import VectorParams


@dataclass
class FP8PackReq:
    xVec: list[int]                    # 16 BF16 bit patterns (uint16)
    expShift: int = 0                  # int8 (post engine-side clamp)


@dataclass
class FP8PackResp:
    result: list[int]                  # 16 UInt16 slots; low byte = byte 2j


class FP8Pack:
    """Mirror of `class FP8Pack` in
    src/main/scala/atlas/vector/laneBoxes/FP8Pack.scala.

    Visible latency = 1 (the `respValidReg` register on the second
    pulse). The class is stateful: `_phase`, `_low_half`,
    `_resp_valid`, and `_resp_bits` mirror the four Scala registers.
    """

    LATENCIES: dict[str, int] = {"fp8pack": 1}

    def __init__(self, p: VectorParams):
        self.p = p
        # `phase = RegInit(false.B)` — false → first half, true → second.
        self._phase: bool = False
        # `lowHalf = Reg(Vec(numLanes, UInt(8.W)))` — 16 buffered FP8 bytes.
        self._low_half: list[int] = [0] * p.num_lanes
        # `respValidReg = RegInit(false.B)`.
        self._resp_valid: bool = False
        # `respBitsReg = Reg(Vec(numLanes, UInt(16.W)))`.
        self._resp_bits: list[int] = [0] * p.num_lanes

    def reset(self) -> None:
        self._phase = False
        self._low_half = [0] * self.p.num_lanes
        self._resp_valid = False
        self._resp_bits = [0] * self.p.num_lanes

    def convert_row(self, req: FP8PackReq) -> list[int]:
        """Pure per-lane BF16 → E4M3 byte conversion for one row.
        No latching, no phase, no output register. Mirrors
        `FP8Pack.scala:42-98` (the combinational `fp8Bytes` wire)."""
        n = self.p.num_lanes
        if len(req.xVec) != n:
            raise ValueError(f"xVec must have {n} lanes, got {len(req.xVec)}")
        return [fp8.bf16_to_e4m3_byte(x, req.expShift) for x in req.xVec]

    def pack_two_rows(
        self, req_low: FP8PackReq, req_high: FP8PackReq
    ) -> FP8PackResp:
        """Pure 2-row pack helper: convert both rows and assemble the
        16 UInt16 output slots, bypassing the latch entirely. Useful
        for tests that want to verify the conversion + packing math
        without driving the cycle-accurate state machine."""
        low_bytes = self.convert_row(req_low)
        high_bytes = self.convert_row(req_high)
        return FP8PackResp(result=self._assemble(low_bytes, high_bytes))

    def peek_resp(self) -> Optional[FP8PackResp]:
        """Return the current registered output without advancing the
        cycle. Mirrors a combinational read of `io.resp.valid` /
        `io.resp.bits.result`."""
        if not self._resp_valid:
            return None
        return FP8PackResp(result=list(self._resp_bits))

    def step(
        self, op_name: str, req: Optional[FP8PackReq]
    ) -> Optional[FP8PackResp]:
        """Cycle-accurate step. Returns whatever `io.resp` shows at
        the START of this cycle (i.e. the registered output that was
        latched at the end of the previous cycle).

        Side effect: at the END of the cycle, the state machine
        advances based on `req`:
          - `req is None` (req.valid=false): clear `respValidReg`,
            preserve `phase` and `lowHalf` across the idle.
          - `req` and `not phase`: convert row, latch into `lowHalf`,
            flip `phase` to true, clear `respValidReg`.
          - `req` and `phase`: convert row, pack both halves into
            `respBitsReg`, set `respValidReg=true`, flip `phase`
            back to false.
        """
        if op_name != "fp8pack":
            raise KeyError(
                f"FP8Pack has no op {op_name!r}; valid: {sorted(self.LATENCIES)}"
            )

        # Output reflects the PREVIOUS cycle's register state.
        output = (
            FP8PackResp(result=list(self._resp_bits))
            if self._resp_valid
            else None
        )

        # End-of-cycle state update.
        if req is None:
            # `}.otherwise { respValidReg := false.B }` — phase and
            # lowHalf are NOT touched, so a late second pulse still
            # finds the correct first-half buffer.
            self._resp_valid = False
        else:
            fp8_bytes = self.convert_row(req)
            if not self._phase:
                self._low_half = fp8_bytes
                self._phase = True
                self._resp_valid = False
            else:
                self._phase = False
                self._resp_valid = True
                self._resp_bits = self._assemble(self._low_half, fp8_bytes)

        return output

    def _assemble(self, low_bytes: list[int], high_bytes: list[int]) -> list[int]:
        """Build the 16 UInt16 output slots from 16 low-half + 16
        high-half FP8 bytes. Mirrors `FP8Pack.scala:122-125` exactly."""
        n = self.p.num_lanes
        half = n // 2
        out = [0] * n
        for j in range(half):
            out[j] = ((low_bytes[2 * j + 1] & 0xFF) << 8) | (low_bytes[2 * j] & 0xFF)
            out[j + half] = (
                ((high_bytes[2 * j + 1] & 0xFF) << 8) | (high_bytes[2 * j] & 0xFF)
            )
        return out

File: src/main/scala/atlas/vector/laneBoxes/FP8Pack.scala
// ============================================================================
// FP8Pack.scala — Vector BF16-to-FP8 conversion with 2→1 packing.
//
// Two consecutive BF16 input rows (16 lanes each) are converted to FP8 and
// packed into a single 256-bit output row (32 FP8 bytes).
//
// Timing: accepts 2 req valids, produces 1 resp valid.
//   req valid 0 → convert 16 BF16 → 16 FP8, buffer in low half
//   req valid 1 → convert 16 BF16 → 16 FP8, pack into high half, output
// ============================================================================

package atlas.vector

import chisel3._
import chisel3.util._
import sp26FPUnits._
import sp26FPUnits.hardfloat._
import sp26FPUnits.hardfloat.consts._

class FP8Req(wordWidth: Int, numLanes: Int) extends Bundle {
  val expShift = SInt(8.W)
  val xVec     = Vec(numLanes, UInt(wordWidth.W))
}

class FP8Resp(wordWidth: Int, numLanes: Int) extends Bundle {
  val result = Vec(numLanes, UInt(wordWidth.W))
}

class FP8Pack(BF16T: AtlasFPType, numLanes: Int = 16)
    extends Module
    with HasPipelineParams {

  val io = IO(new Bundle {
    val req  = Flipped(Valid(new FP8Req(BF16T.wordWidth, numLanes)))
    val resp = Valid(new FP8Resp(BF16T.wordWidth, numLanes))
  })

  // E4M3 constants
  val E4M3_MAX_POS = "h7e".U(8.W)
  val E4M3_MAX_NEG = "hfe".U(8.W)

  // ── Per-lane BF16 → FP8 conversion (combinational) ──────────────
  val fp8Bytes = Wire(Vec(numLanes, UInt(8.W)))

  for (i <- 0 until numLanes) {
    val bf16 = io.req.bits.xVec(i)

    val sign   = bf16(15)
    val expBF  = bf16(14, 7)
    val mantBF = bf16(6, 0)

    val isZero = (expBF === 0.U) && (mantBF === 0.U)
    val isSub  = (expBF === 0.U) && (mantBF =/= 0.U)
    val isInf  = (expBF === "hff".U) && (mantBF === 0.U)
    val isNaN  = (expBF === "hff".U) && (mantBF =/= 0.U)

    val unbExp       = expBF.zext - 127.S(10.W)
    val scaleExpWide = io.req.bits.expShift.pad(10)
    val expAdjusted  = unbExp - scaleExpWide

    val mant8      = Cat(1.U(1.W), mantBF)
    val trunc      = Cat(0.U(1.W), mant8(7, 4))
    val guard      = mant8(3)
    val sticky     = mant8(2, 0).orR
    val lsb        = trunc(0)
    val inc        = guard && (sticky || lsb)
    val roundedSig = trunc + inc
    val normCarry  = (roundedSig === 16.U(5.W))

    val finalExpAdjusted = Wire(SInt(10.W))
    finalExpAdjusted := Mux(normCarry, expAdjusted + 1.S, expAdjusted)

    val mantFP8 = Wire(UInt(3.W))
    mantFP8 := Mux(normCarry, 0.U, (roundedSig - 8.U)(2, 0))

    val fp8 = Wire(UInt(8.W))
    fp8 := 0.U
    when(isZero || isSub || isNaN) {
      fp8 := 0.U
    }.elsewhen(isInf) {
      fp8 := Mux(sign.asBool, E4M3_MAX_NEG, E4M3_MAX_POS)
    }.otherwise {
      when(finalExpAdjusted > 8.S) {
        fp8 := Mux(sign.asBool, E4M3_MAX_NEG, E4M3_MAX_POS)
      }.elsewhen(finalExpAdjusted >= (-6).S) {
        val expFP8 = (finalExpAdjusted + 7.S).asUInt(3, 0)
        val packed = Cat(sign, expFP8, mantFP8)
        fp8 := Mux(
          (finalExpAdjusted === 8.S) && (mantFP8 === 7.U),
          Mux(sign.asBool, E4M3_MAX_NEG, E4M3_MAX_POS),
          packed
        )
      }.otherwise {
        fp8 := 0.U
      }
    }
    fp8Bytes(i) := fp8
  }

  // ── 2→1 accumulation ───────────────────────────────────────────
  val phase        = RegInit(false.B)              // false = first half, true = second
  val lowHalf      = Reg(Vec(numLanes, UInt(8.W))) // buffered first 16 FP8 bytes
  val respValidReg = RegInit(false.B)
  val respBitsReg  = Reg(Vec(numLanes, UInt(BF16T.wordWidth.W)))

  io.resp.valid := respValidReg
  io.resp.bits.result := respBitsReg

  when(io.req.valid) {
    when(!phase) {
      // First input: buffer 16 FP8 bytes, no output yet
      lowHalf := fp8Bytes
      phase   := true.B
      respValidReg := false.B
    }.otherwise {
      // Second input: pack both halves into one 256-bit row
      // result(i) = bits[16i+15 : 16i]. Each 16-bit slot holds two FP8 bytes.
      // Low half (bytes 0..15) → result(0)..result(7)
      // High half (bytes 16..31) → result(8)..result(15)
      phase        := false.B
      respValidReg := true.B
      for (j <- 0 until numLanes / 2) {
        respBitsReg(j)              := Cat(lowHalf(2 * j + 1),  lowHalf(2 * j))
        respBitsReg(j + numLanes/2) := Cat(fp8Bytes(2 * j + 1), fp8Bytes(2 * j))
      }
    }
  }.otherwise {
    respValidReg := false.B
  }
}
